rosbag2_storage_mcap/src/mcap_storage.cpp

```diff
@@ -567,90 +645,28 @@ void MCAPStorage::reset_iterator()
-                            storage_filter_.exclude_topics.end(), topic_string);
-        if (it != storage_filter_.exclude_topics.end()) {
-          return false;
-        }
-      }
-    }
-
-#ifdef ROSBAG2_STORAGE_MCAP_HAS_STORAGE_FILTER_TOPIC_REGEX
-    if (!storage_filter_.regex_to_exclude.empty()) {
-      std::smatch m;
-      std::regex re(storage_filter_.regex_to_exclude);
-
-      if (std::regex_match(topic_string, m, re)) {
-        return false;
-      }
-    }
-#endif
-
-    if (is_service_event_topic) {
-      if (!storage_filter_.services_events.empty()) {
-        auto it = std::find(storage_filter_.services_events.begin(),
-                            storage_filter_.services_events.end(), topic_string);
-        if (it != storage_filter_.services_events.end()) {
-          return true;
-        } else {
-#ifdef ROSBAG2_STORAGE_MCAP_HAS_STORAGE_FILTER_TOPIC_REGEX
-          // If regex isn't set, service name must be in service event topic list
-          if (storage_filter_.regex.empty()) {
-            return false;
-          }
-#else
-          return false;
-#endif
-        }
-      }
-    } else {
-      if (!storage_filter_.topics.empty()) {
-        auto it =
-          std::find(storage_filter_.topics.begin(), storage_filter_.topics.end(), topic_string);
-        if (it != storage_filter_.topics.end()) {
-          return true;
-        } else {
-#ifdef ROSBAG2_STORAGE_MCAP_HAS_STORAGE_FILTER_TOPIC_REGEX
-          // If regex isn't set, topic name must be in topic topic list
-          if (storage_filter_.regex.empty()) {
-            return false;
-          }
-#else
-          return false;
-#endif
-        }
-      }
-    }
+  auto topic_filter = [this](std::string_view topic) {
+    bool topic_a_service_event = is_topic_name_a_service_event(topic);
 
-#ifdef ROSBAG2_STORAGE_MCAP_HAS_STORAGE_FILTER_TOPIC_REGEX
-    // If a regex is defined, the topic name or service event name must match the regex.
-    if (!storage_filter_.regex.empty()) {
-      std::smatch m;
-      std::regex re(storage_filter_.regex);
+    const auto & include_list =
+      topic_a_service_event ? storage_filter_.services_events : storage_filter_.topics;
 
-      if (!std::regex_match(topic_string, m, re)) {
+    const auto & exclude_list = topic_a_service_event ? storage_filter_.exclude_service_events
+                                                      : storage_filter_.exclude_topics;
+    // if topic found in include list or regex
+    if (is_topic_selected_by_white_list_or_regex(topic, include_list, storage_filter_.regex)) {
+      // if topic found in exclude list or regex_to_exclude
+      if (is_topic_in_black_list_or_exclude_regex(topic, exclude_list,
+                                                  storage_filter_.regex_to_exclude)) {
         return false;
+      } else {
+        return true;
       }
+    } else {  // topic not found in include list or regex
+      return false;
     }
```

Comment:
I think is_topic_in_black_list_or_exclude_regex() should be checked at the first.    
Users might use regex to include topics, but want to remove some using an exclude_list. If we first execute is_topic_selected_by_white_list_or_regex(), we still need to execute is_topic_in_black_list_or_exclude_regex().

```suggestion
    // if topic not found in exclude list or regex_to_exclude
    if (!is_topic_in_black_list_or_exclude_regex(topic, exclude_list,
                                                 storage_filter_.regex_to_exclude))
    {
      // if topic found in include list or regex
      if (is_topic_selected_by_white_list_or_regex(topic, include_list, storage_filter_.regex)) {
        return true;
      }
    }

    return false;
```

Reply:
Good catch. Make sense to me.